### PR TITLE
New VersionFile module

### PR DESCRIPTION
**Pull request by heksesang**
A module to help set release version and update to next snapshot when publishing.

### CODE REVIEW

**Comment by heksesang** [comment]
I'll try add the suggested changes. :)

Is there any documentation anywhere on what the different `T` types are and how they work?

And about the name, "versionFile" is the only of your suggestions that I think makes it clear that it's not actually doing any release action. Though, if we also add support for using the git tag instead of a version file, is that name a bit misleading again?

**Comment by lefou** [comment]
The different Task types are documented here: http://www.lihaoyi.com/mill/page/tasks.html

Yeah, I guess "versionFile" for the current PR is the best. If you want to integrate the git idea it won't fit, I agree. But the git idea was only a pointer or suggestion. Do what you want.

**Comment by heksesang** [comment]
I'll change it to versionFile and not add the git idea in this PR. The git idea should probably be made into some module, but maybe it should be its own module? That would make naming easier.

File: contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+trait VersionFileModule extends Module {
+
+  implicit val wd = os.pwd
+
+  def versionFileName = "version"
+  def versionFileDirectory = wd
+  def versionFilePath = versionFileDirectory / versionFileName
+
+  def currentVersion = T.input { Version.of(os.read(versionFilePath)) }
```
Comment (lefou) [inline contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala]:
```suggestion
  /** The file containing the current version. */
  def versionFile: Source = T.source(millSourcePath / "version")
  /** The current version. */
  def currentVersion: T[Version] = T { Version.of(os.read(versionFilePath().path)) }
```

That way, we avoid reading the file if it hasn't changed. It looks also a bit more idiomatic (if that is even a thing).

Comment (heksesang) [reply contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala]: Won't `millSourcePath / "version"` prevent one from having the version file at root level of the repository?

Comment (lefou) [reply contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala]: Yeah. It's an oversight from me. But `wd` isn't mill API. You can either override `millSourcePath` or you can use the value `millOuterCtx.millSourcePath`.

File: docs/pages/9 - Contrib Modules.md [hunk 2]
```diff
@@ -845,3 +844,118 @@ These imports will always be added to every template.  You don't need to list th
...
+}
+```
+
+### Configure the version file
+This module reads and updates a version file. The default is the file `version` at the project root.
+
+If you prefer something else than the default, you can override:
+* `VersionFileModule.versionFileName` which defaults to `version`
+* `VersionFileModule.versionFileDirectory` which defaults to `os.pwd`
+* `VersionFileModule.versionFilePath` which defaults to `versionFileDirectory / versionFileName`
```
Comment (heksesang) [inline docs/pages/9 - Contrib Modules.md]:
```suggestion
This module reads and updates a version file. The default is the file `version` at the module source root. If you prefer something other than the default, you can override `VersionFileModule.versionFile`.
```

File: contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  def releaseVersion = T { currentVersion().asRelease }
+  def nextVersion(bump: String) = T.command { releaseVersion().bump(bump) }
```
Comment (heksesang) [inline contrib/versionfile/src/mill/contrib/versionfile/VersionFileModule.scala]:
```suggestion
  /** The release version. */
  def releaseVersion = T { currentVersion().asRelease }
  /** The next snapshot version. */
  def nextVersion(bump: String) = T.command { releaseVersion().bump(bump) }
```
Might as well add comments to these two too. :)

**Comment by lefou** [comment]
Thank you!

**Pull request merged**